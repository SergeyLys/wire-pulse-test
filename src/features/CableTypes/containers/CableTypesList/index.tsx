export {default as CableTypesList} from './CableTypesList';