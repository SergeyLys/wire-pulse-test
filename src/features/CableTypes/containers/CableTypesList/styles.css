.list-controls {
  z-index: 1;
  position: sticky;
  top: 0;
  width: 100%;
  background-color: white;
  padding: 15px 0;
}