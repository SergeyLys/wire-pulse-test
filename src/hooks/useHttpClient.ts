import HttpClient from "../libs/httpClient";

type UseHttpClient = {
  get: <T>(url: string, options?: Record<string, any>) => Promise<T>;
}

const client = new HttpClient();
client.baseUrl = 'http://localhost:5174';

const useHttpClient = (): UseHttpClient => {
  return {
    get: <T>(url: string, options?: Record<string, any>) => client.get<T>(url, options).then(data => data).catch(error => {
      throw error.message;
    })
  }
}

export default useHttpClient;