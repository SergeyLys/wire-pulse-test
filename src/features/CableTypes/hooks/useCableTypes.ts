import { useState, useEffect, useCallback } from "react";
import { CableType, CableTypeResponse } from "../../../../shared/types";
import { setQueryParam } from "../../../libs/queryParams";
import useHttpClient from "../../../hooks/useHttpClient";

type UseCableTypes = {
  isLoading: boolean; 
  error: string | null; 
  list: CableType[]; 
  total: number;
}

const useCableTypes = (limit: number, after: number, searchValue: string): UseCableTypes => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [list, setList] = useState<CableType[]>([]);
  const [total, setTotal] = useState<number>(0);
  const { get } = useHttpClient();

  const sendQuery = useCallback(async (url: string) => get<CableTypeResponse>(url), []);

  useEffect(() => {
    const url = setQueryParam('/cable/type', {limit, after, searchValue});
    setIsLoading(true);
    setError(null);
      
    sendQuery(url)
      .then(res => {
        setList(res!.entities);
        setTotal(res!.total);
      })
      .catch(err => {
        setError(err);
      })
      .finally(() => {
        setIsLoading(false);
      });
  }, [limit, after, searchValue, sendQuery]);

  return { isLoading, error, list, total };
}

export {type CableType};
export default useCableTypes;