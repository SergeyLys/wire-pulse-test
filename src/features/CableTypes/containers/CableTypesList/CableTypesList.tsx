import { useMemo, useState } from "react";
import useCableTypes from "../../hooks/useCableTypes";
import { ListView, ColumnType, Pagination, Input } from "../../../../ui";
import debounce from "../../../../libs/debounce";
import "./styles.css";

const columns: ColumnType[] = [
  {
    name: 'Catid',
    styles: {width: 100},
    dataRef: 'catid',
    key: '0',
  },
  {
    name: 'Manufacturer',
    styles: {width: '70%'},
    dataRef: 'manufacturer',
    key: '1',
  },
  {
    name: 'Diameter',
    dataRef: 'diameter',
    key: '2',
    styles: {width: 120},
  }
];

const pageSizeOptions = [{id: 10, value: '10'}, {id: 20, value: '20'}, {id: 100, value: '100'}];

const DEFAULT_PAGE_SIZE = 10;
const DEFAULT_CURRENT_PAGE = 1;

const CableTypesList = () => {
  const [pagination, setPagination] = useState({after: 0, currentPage: DEFAULT_CURRENT_PAGE, pageSize: DEFAULT_PAGE_SIZE});
  const [searchValue, setSearchValue] = useState('');
  const { list, isLoading, error, total } = useCableTypes(pagination.pageSize, pagination.after, searchValue);

  const handleSearch = useMemo(() => debounce((value: string) => {
    setSearchValue(value);
  }, 500), []);

  const handleChangePage = useMemo(() => debounce((page: number, pageSize: number) => {
    setPagination(() => ({after: (page - 1) * pageSize, currentPage: page, pageSize}));
  }, 500), []);

  const normalizedData = useMemo(() => list.map(item => ({
    key: item.catid,
    catid: item.catid,
    manufacturer: item.manufacturer.name,
    diameter: `${item.diameter.actual.value} ${item.diameter.actual.unit}`
  })), [list]);
  
  return (
    <div className="relative">
      <div className="list-controls">
        <div className="flex wrap">
          <div className="flex flex-column flex-grow-1 justify-center">
            <Input testId="cable-types-list-search-input" value="" onChange={handleSearch} />
          </div>
          <Pagination
            total={total} 
            defaultPageSize={DEFAULT_PAGE_SIZE}
            current={pagination.currentPage}
            onChange={handleChangePage}
          >
            <Pagination.PageSizeSelector styles={{marginRight: 15}} pageSizeOptions={pageSizeOptions} />
            <Pagination.Arrow direction="left" />
            <Pagination.Current />
            <Pagination.Arrow direction="right" />
          </Pagination>
        </div>
      </div>
      <ListView
        headerStyles={{
          position: 'sticky',
          top: '91px'
        }}
        columns={columns}
        data={normalizedData}
        error={error}
        isLoading={isLoading}
      />
    </div>
  )
}

export {DEFAULT_PAGE_SIZE, DEFAULT_CURRENT_PAGE};
export default CableTypesList;