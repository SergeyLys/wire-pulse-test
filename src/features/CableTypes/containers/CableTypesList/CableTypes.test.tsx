import renderer from 'react-test-renderer';
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react';
import CableTypesList, {DEFAULT_PAGE_SIZE} from './CableTypesList';
import * as useCableTypes from '../../hooks/useCableTypes';

const mockEntity = {
  "id": "5f3bc9e2502422053e08f9f1",
  "identifier": "10-al-1c-trxple",
  "catid": 1622475,
  "diameter": {
    "published": {
      "value": 22.43,
      "unit": "mAh"
    },
    "actual": {
      "value": 22.43,
      "unit": "mAh"
    }
  },
  "conductor": {
    "number": 0,
    "size": {
      "value": 22.43,
      "unit": "mAh"
    }
  },
  "insulation": {
    "type": "string",
    "shield": "string",
    "jacket": "string",
    "thickness": {
      "value": 22.43,
      "unit": "mAh"
    }
  },
  "material": {
    "aluminum": 0,
    "copper": 0,
    "weight": {
      "net": {
        "value": 22.43,
        "unit": "mAh"
      },
      "calculated": {
        "value": 22.43,
        "unit": "mAh"
      }
    }
  },
  "currentPrice": {
    "value": 22.43,
    "unit": "USD"
  },
  "voltage": {
    "value": 22.43,
    "unit": "mAh"
  },
  "rotationFrequency": {
    "value": 22.43,
    "unit": "mAh"
  },
  "manufacturer": {
    "id": "5f3bc9e2502422053e08f9f1",
    "name": "Kerite"
  },
  "properties": [
    {
      "name": "manufacturedBy",
      "value": {
        "string": "string value",
        "number": 1234.56
      }
    }
  ],
  "customer": {
    "id": "5f3bc9e2502422053e08f9f1",
    "code": "bge"
  },
  "metadata": {
    "created": new Date("2020-10-13T21:31:51.259Z"),
    "modified": new Date("2020-10-13T21:31:51.259Z"),
    "user": {
      "id": "5f3bc9e2502422053e08f9f1",
      "username": "test@reelsense.io"
    }
  }
};

jest.mock("../../hooks/useCableTypes", () => jest.fn(() => {
  return {
    list: [mockEntity],
    isLoading: false,
    error: null,
    total: 1
  }
}));

const setupComponent = () => <CableTypesList />

describe('CableTypesList component', () => {
  jest.useFakeTimers();

  beforeEach(() => {
    jest.clearAllMocks();
  })

  it('renders correctly', () => {
    const tree = renderer.create(
      setupComponent()
    ).toJSON();
    render(setupComponent());

    expect(tree).toMatchSnapshot();
    expect(screen.getByText(mockEntity.catid)).toBeInTheDocument();
  });

  it('should react on input change after debounce timer', async () => {
    const spy = jest.spyOn(useCableTypes, 'default').mockImplementation(() => {
      return {
        list: [],
        isLoading: false,
        error: null,
        total: 0
      }
    });

    render(setupComponent());

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      fireEvent.input(screen.getByTestId('cable-types-list-search-input'), {
        target: {
          value: 'test input'
        }
      });
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(DEFAULT_PAGE_SIZE, 0, '');
      jest.advanceTimersByTime(500);
    });

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith(DEFAULT_PAGE_SIZE, 0, 'test input');
  });

  it('should react on pagination changes after debounce timer', async () => {
    const spy = jest.spyOn(useCableTypes, 'default').mockImplementation(() => {
      return {
        list: [],
        isLoading: false,
        error: null,
        total: 100
      }
    });

    render(setupComponent());

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      fireEvent.click(screen.getByTestId('pagination-arrow-next'));
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(DEFAULT_PAGE_SIZE, 0, '');
      jest.advanceTimersByTime(500);
    });

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith(DEFAULT_PAGE_SIZE, 10, '');

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      fireEvent.click(screen.getByTestId('pagination-arrow-prev'));
      jest.advanceTimersByTime(500);
    });

    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy).toHaveBeenCalledWith(DEFAULT_PAGE_SIZE, 0, '');

    fireEvent.click(screen.getByTestId('pagination-page-size-selector'));
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      fireEvent.click(screen.getAllByTestId('pagination-page-size-selector-item')[1]);
      jest.advanceTimersByTime(500);
    });

    expect(spy).toHaveBeenCalledTimes(4);
    expect(spy).toHaveBeenCalledWith(20, 0, '');
  });
});
