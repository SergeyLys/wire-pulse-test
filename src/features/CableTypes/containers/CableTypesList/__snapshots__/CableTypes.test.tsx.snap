// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CableTypesList component renders correctly 1`] = `
<div
  className="relative"
>
  <div
    className="list-controls"
  >
    <div
      className="flex wrap"
    >
      <div
        className="flex flex-column flex-grow-1 justify-center"
      >
        <input
          className="input"
          data-testid="cable-types-list-search-input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="flex pagination-container align-center"
      >
        <div
          style={
            Object {
              "marginRight": 15,
            }
          }
        >
          <div
            className="dropdown"
          >
            <div
              data-testid="pagination-page-size-selector"
              onClick={[Function]}
            >
              <div
                className="dropdown-button "
              >
                10
              </div>
            </div>
          </div>
        </div>
        <div
          className="prev arrow"
          data-testid="pagination-arrow-prev"
          onClick={[Function]}
        >
          <svg
            height="800px"
            version="1.1"
            viewBox="0 0 34.075 34.075"
            width="800px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <g>
                <path
                  d="M24.57,34.075c-0.505,0-1.011-0.191-1.396-0.577L8.11,18.432c-0.771-0.771-0.771-2.019,0-2.79 L23.174,0.578c0.771-0.771,2.02-0.771,2.791,0s0.771,2.02,0,2.79l-13.67,13.669l13.67,13.669c0.771,0.771,0.771,2.021,0,2.792 C25.58,33.883,25.075,34.075,24.57,34.075z"
                  style={
                    Object {
                      "fill": "#010002",
                    }
                  }
                />
              </g>
            </g>
          </svg>
        </div>
        <div
          className="current"
          data-testid="pagination-current-page"
        >
          1
           of 
          1
        </div>
        <div
          className="next arrow"
          data-testid="pagination-arrow-next"
          onClick={[Function]}
        >
          <svg
            height="800px"
            version="1.1"
            viewBox="0 0 185.343 185.343"
            width="800px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <g>
                <path
                  d="M51.707,185.343c-2.741,0-5.493-1.044-7.593-3.149c-4.194-4.194-4.194-10.981,0-15.175 l74.352-74.347L44.114,18.32c-4.194-4.194-4.194-10.987,0-15.175c4.194-4.194,10.987-4.194,15.18,0l81.934,81.934 c4.194,4.194,4.194,10.987,0,15.175l-81.934,81.939C57.201,184.293,54.454,185.343,51.707,185.343z"
                  style={
                    Object {
                      "fill": "#010002",
                    }
                  }
                />
              </g>
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="list-container"
  >
    <div
      className="list-header flex"
      style={
        Object {
          "position": "sticky",
          "top": "91px",
        }
      }
    >
      <div
        style={
          Object {
            "width": 100,
          }
        }
      >
        Catid
      </div>
      <div
        style={
          Object {
            "width": "70%",
          }
        }
      >
        Manufacturer
      </div>
      <div
        style={
          Object {
            "width": 120,
          }
        }
      >
        Diameter
      </div>
    </div>
    <div
      className="list-body"
    >
      <div
        className="flex list-body-item"
        data-testid="listview-item"
      >
        <div
          className="list-body-col-item"
          style={
            Object {
              "width": 100,
            }
          }
        >
          1622475
        </div>
        <div
          className="list-body-col-item"
          style={
            Object {
              "width": "70%",
            }
          }
        >
          Kerite
        </div>
        <div
          className="list-body-col-item"
          style={
            Object {
              "width": 120,
            }
          }
        >
          22.43 mAh
        </div>
      </div>
    </div>
  </div>
</div>
`;
